src/basic/fd-util.h

```diff
@@ -35,6 +35,13 @@ int close_nointr(int fd);
 int safe_close(int fd);
 void safe_close_pair(int p[]);
 
+static inline int safe_close_above_stdio(int fd) {
+        if (fd < 3) /* keep stdin/stdout/stderr open + propagate closed as closed. */
```

Comment:
Some callers would benefit from -1 being returned unconditionally, and others wouldn't care. Maybe this should be changed?